File: common/mine-with-friends-board/src/game/game_move.rs
use crate::nfts::instances::powerplant::PlantType;
use crate::nfts::NftPtr;
use crate::tokens::token_swap::TradingPairID;
use crate::util::Currency;
use crate::{entity::EntityID, util::Price};
use crate::{sanitize, MoveEnvelope};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

/// Grab-Bag Enum of all moves
///
/// N.B. we do the enum-of-struct-variant pattern to make serialization/schemas
/// nicer.
#[derive(Debug, Serialize, Deserialize, JsonSchema, Eq, PartialEq, Clone)]
#[serde(rename_all = "snake_case")]
pub enum GameMove {
    Heartbeat(Heartbeat),
    /// # Trade Coins
    Trade(Trade),
    /// # Buy NFTs
    PurchaseNFT(PurchaseNFT),
    /// # Sell NFTs
    ListNFTForSale(ListNFTForSale),
    /// # Send Coins
    SendTokens(SendTokens),
    /// # Remove Tokens
    RemoveTokens(RemoveTokens),
    /// # Send a logged Chat Message to All Players
    Chat(Chat),
    /// # Mint Power Plant NFT
    MintPowerPlant(MintPowerPlant),
    /// # Purchase Materials, then Mint Power Plant NFT
    SuperMintPowerPlant(MintPowerPlant),
}

// Convenience to marshall a move into a GameMove
macro_rules! derive_from {
    ($y:ident) => {
        impl From<$y> for GameMove {
            fn from(i: $y) -> Self {
                GameMove::$y(i)
            }
        }
    };
}
derive_from!(Heartbeat);
derive_from!(Trade);
derive_from!(MintPowerPlant);
derive_from!(PurchaseNFT);
derive_from!(ListNFTForSale);
derive_from!(SendTokens);
derive_from!(RemoveTokens);
derive_from!(Chat);

#[derive(Debug, Eq, PartialEq, Serialize, Deserialize, JsonSchema, Clone)]
pub struct Heartbeat();

#[derive(Debug, Eq, PartialEq, Serialize, Deserialize, JsonSchema, Clone)]
pub struct Trade {
    pub pair: TradingPairID,
    pub amount_a: u128,
    pub amount_b: u128,
    pub sell: bool,
    #[serde(default)]
    pub cap: Option<u128>,
}

#[derive(Debug, Eq, PartialEq, Serialize, Deserialize, JsonSchema, Clone)]
pub struct MintPowerPlant {
    /// Size of the power plant
    pub scale: u64,
    pub location: (i64, i64),
    pub plant_type: PlantType,
}

#[derive(Debug, Eq, PartialEq, Serialize, Deserialize, JsonSchema, Clone)]
pub struct PurchaseNFT {
    pub nft_id: NftPtr,
    pub limit_price: Price,
    pub currency: Currency,
}

#[derive(Debug, Eq, PartialEq, Serialize, Deserialize, JsonSchema, Clone)]
pub struct ListNFTForSale {
    pub nft_id: NftPtr,
    pub price: Price,
    pub currency: Currency,
}

#[derive(Debug, Eq, PartialEq, Serialize, Deserialize, JsonSchema, Clone)]
pub struct SendTokens {
    pub to: EntityID,
    pub amount: Price,
    pub currency: Currency,
}

#[derive(Debug, Eq, PartialEq, Serialize, Deserialize, JsonSchema, Clone)]
pub struct RemoveTokens {
    pub nft_id: NftPtr,
    pub amount: Price,
    pub currency: Currency,
    // do we need time here?
}

#[derive(Debug, Eq, PartialEq, Serialize, Deserialize, JsonSchema, Clone)]
pub struct Chat(pub String);

impl MoveEnvelope {
    pub fn create<G: Into<GameMove>>(
        g: G,
        sequence: u64,
        _sig: String,
        _from: EntityID,
        time: u64,
    ) -> Self {
        MoveEnvelope {
            d: sanitize::Unsanitized(g.into()),
            sequence,
            time_millis: time,
        }
    }
}

File: ux/src/Types/GameMove.ts
/**
 * Grab-Bag Enum of all moves
 *
 * N.B. we do the enum-of-struct-variant pattern to make serialization/schemas nicer.
 */
export type GameMove =
  | {
    heartbeat: Heartbeat
  }
  | TradeCoins
  | BuyNFTs
  | SellNFTs
  | SendCoins
  | RemoveTokens
  | SendALoggedChatMessageToAllPlayers
  | MintPowerPlantNFT
  | PurchaseMaterialsThenMintPowerPlantNFT
export type Heartbeat = []
export type TradingPairID = string
/**
* an EntityID is just a "pointer" we assign to all different types of things in our game, e.g. - Users - Token Contracts - NFTs - etc
*
* EntityIDs are global and unique within the game state
*/
export type EntityID = string
export type Chat = string
export type PlantType = "Solar" | "Hydro" | "Flare"

export interface TradeCoins {
  trade: Trade
}
export interface Trade {
  amount_a: number
  amount_b: number
  cap?: number | null
  pair: TradingPairID
  sell: boolean
  [k: string]: unknown
}
export interface BuyNFTs {
  purchase_n_f_t: PurchaseNFT
}
export interface PurchaseNFT {
  currency: EntityID
  limit_price: number
  nft_id: EntityID
  [k: string]: unknown
}
export interface SellNFTs {
  list_n_f_t_for_sale: ListNFTForSale
}
export interface ListNFTForSale {
  currency: EntityID
  nft_id: EntityID
  price: number
  [k: string]: unknown
}
export interface SendCoins {
  send_tokens: SendTokens
}
export interface SendTokens {
  amount: number
  currency: EntityID
  to: EntityID
  [k: string]: unknown
}
export interface RemoveTokens {
  remove_tokens: RemoveTokens1
}
export interface RemoveTokens1 {
  amount: number
  currency: EntityID
  nft_id: EntityID
  [k: string]: unknown
}
export interface SendALoggedChatMessageToAllPlayers {
  chat: Chat
}
export interface MintPowerPlantNFT {
  mint_power_plant: MintPowerPlant
}
export interface MintPowerPlant {
  location: [number, number]
  plant_type: PlantType
  /**
   * Size of the power plant
   */
  scale: number
  [k: string]: unknown
}
export interface PurchaseMaterialsThenMintPowerPlantNFT {
  super_mint_power_plant: MintPowerPlant
}
